src/sources/exec.rs

```diff
@@ -0,0 +1,830 @@
+            error!(message = "Failed to forward events; downstream is closed.");
+        });
+}
+
+fn build_command(config: &ExecConfig) -> Command {
+    let command = &config.command[0];
+
+    let mut command = Command::new(command);
+
+    if config.command.len() > 1 {
+        command.args(&config.command[1..]);
+    };
+
+    command.kill_on_drop(true);
+
+    // Explicitly set the current dir if needed
+    if let Some(current_dir) = &config.working_directory {
+        command.current_dir(current_dir);
+    }
+
+    // Pipe our stdout/stderr to the process to we inherit it's output
+    command.stdout(std::process::Stdio::piped());
+    command.stderr(std::process::Stdio::piped());
+
+    command
+}
+
+fn create_event(
+    config: &ExecConfig,
+    hostname: &Option<String>,
+    line: Bytes,
+    data_stream: &Option<String>,
+    pid: Option<u32>,
+    exit_status: Option<i32>,
+    exec_duration_millis: &Option<u128>,
+) -> Event {
+    emit!(ExecEventReceived {
+        command: config.command_line().as_str(),
+        byte_size: line.len(),
+    });
+    let mut log_event = LogEvent::default();
+
+    // Add message
+    log_event.insert(log_schema().message_key(), line);
+
+    // Add timestamp
+    log_event.insert(log_schema().timestamp_key(), Utc::now());
+
+    // Add source type
+    log_event.insert(log_schema().source_type_key(), Bytes::from(EXEC));
+
+    // Add data stream of stdin or stderr (if needed)
+    if let Some(data_stream) = data_stream {
+        log_event.insert(DATA_STREAM_KEY, data_stream.clone());
+    }
+
+    // Add pid (if needed)
+    if let Some(pid) = pid {
+        log_event.insert(PID_KEY, pid as i64);
+    }
+
+    // Add exit status (if needed)
+    if let Some(exit_status) = exit_status {
+        log_event.insert(EXIT_STATUS_KEY, exit_status as i64);
+    }
+
+    // Add exec duration millis (if needed)
+    if let Some(exec_duration_millis) = exec_duration_millis {
+        log_event.insert(EXEC_DURATION_MILLIS_KEY, *exec_duration_millis as i64);
+    }
+
+    // Add hostname (if needed)
+    if let Some(hostname) = hostname {
+        log_event.insert(log_schema().host_key(), hostname.clone());
+    }
+
+    // Add command
+    log_event.insert(COMMAND_KEY, config.command.clone());
+
+    Event::Log(log_event)
+}
+
+fn spawn_reader_thread<R: 'static + AsyncRead + Unpin + std::marker::Send>(
+    reader: BufReader<R>,
+    shutdown: ShutdownSignal,
+    event_per_line: bool,
+    buf_size: u64,
+    stream: &'static str,
+    sender: Sender<(Bytes, &'static str)>,
+) {
+    // Start the green background thread for collecting
+    Box::pin(tokio::spawn(async move {
+        debug!("Start capturing {} command output.", stream);
+
+        let mut read_buffer: Vec<u8> = Vec::new();
+
+        let reader = reader.allow_read_until(shutdown.clone().map(|_| ()));
+
+        let mut limited_reader = reader.take(buf_size);
+
+        if event_per_line {
```

Comment:
Not a big deal, but it did strike me that maybe the below could be implemented using tokio codecs to unify the two branch arms.

If `event_per_line` is true, it could use https://docs.rs/tokio-util/0.6.5/tokio_util/codec/struct.LinesCodec.html
If `event_per_line` is false, it could use https://docs.rs/tokio-util/0.6.5/tokio_util/codec/struct.BytesCodec.html

Reply:
Cool! I was not aware of that. I'll need at bit more time to look at this tomorrow so I'll first submit the other fixes back to you.

Reply:
Aha, I think `event_per_line = true` should result in the `\r` and `\n` being trimmed.

Reply:
The `LinesCodec` does remove the end of line characters.

Reply:
This looks really good!

I think you could simplify it even further with something like:

```rust
let codec = if event_per_line {
  LinesCodec::new_with_max_length(buf_size)
} else {
  SizedBytesCodec::new_with_max_length(buf_size)
};
```

You might need to `Box` the codec since it is a trait object (`let codec: Box<dyn Codec>`).

That would allow you to DRY up the rest of this since it is identical aside from the codec creation.